src/Microsoft.TestPlatform.CrossPlatEngine/Discovery/DiscovererEnumerator.cs

```diff
@@ -135,93 +135,174 @@ private void LoadTestsFromAnExtension(string extensionAssembly, IEnumerable<stri
+        {
+            discovererType = null;
+
+            // See if discoverer can be instantiated successfully else move next.
+            try
+            {
+                discovererType = discoverer.Value.GetType();
+            }
+            catch (Exception e)
+            {
+                var mesage = string.Format(
+                    CultureInfo.CurrentUICulture,
+                    CrossPlatEngineResources.DiscovererInstantiationException,
+                    e.Message);
+                logger.SendMessage(TestMessageLevel.Warning, mesage);
+                EqtTrace.Error("DiscovererEnumerator.LoadTestsFromAnExtension: {0} ", e);
+
+                return false;
+            }
+
+            return true;
+        }
+
+        private void RecordTotalTestsDiscoveredByCurrentAdapter(long currentTotalTests, LazyExtension<ITestDiscoverer, ITestDiscovererCapabilities> discoverer,
+            ref bool discoverersFromDeprecatedLocations, ref double totalAdaptersUsed)
+        {
+            if (this.discoveryResultCache.TotalDiscoveredTests > currentTotalTests)
+            {
+                var totalDiscoveredTests = this.discoveryResultCache.TotalDiscoveredTests - currentTotalTests;
+                this.requestData.MetricsCollection.Add(
+                    string.Format("{0}.{1}", TelemetryDataConstants.TotalTestsByAdapter,
+                        discoverer.Metadata.DefaultExecutorUri), totalDiscoveredTests);
+                if (!CrossPlatEngine.Constants.DefaultAdapters.Contains(discoverer.Metadata.DefaultExecutorUri.ToString(),
+                    StringComparer.OrdinalIgnoreCase))
                 {
-                    var message = string.Format(
-                        CultureInfo.CurrentUICulture,
-                        CrossPlatEngineResources.ExceptionFromLoadTests,
-                        discovererType.Name,
-                        e.Message);
+                    var discovererLocation = discoverer.Value.GetType().GetTypeInfo().Assembly.GetAssemblyLocation();
 
-                    logger.SendMessage(TestMessageLevel.Error, message);
-                    EqtTrace.Error("DiscovererEnumerator.LoadTestsFromAnExtension: {0} ", e);
+                    discoverersFromDeprecatedLocations |= Path.GetDirectoryName(discovererLocation)
+                        .Equals(CrossPlatEngine.Constants.DefaultAdapterLocation, StringComparison.OrdinalIgnoreCase);
                 }
+
+                totalAdaptersUsed++;
             }
+        }
 
-            // Collecting Total Time Taken by Adapters
-            this.requestData.MetricsCollection.Add(TelemetryDataConstants.TimeTakenInSecByAllAdapters, totalTimeTakenByAdapters);
+        private static void LogWarningOnNoTestsDiscovered(IEnumerable<string> sources, string testCaseFilter, IMessageLogger logger)
+        {
+            var sourcesString = string.Join(" ", sources);
```

Comment:
Should we join strings by \n instead of space?

Reply:
Keeping in one makes easy to copy and run tests from those assemblies again from CLI and currently we are keeping in one line other places.